### ISSUE TITLE
ofi: provider tcp;ofi_rxm is not being picked up

### ISSUE THREAD

**Issue by aingerson** [pmodels/mpich#4188]
When trying to run MPI with tcp;ofi_rxm, I usually set FI_PROVIDER=tcp which automatically picks up tcp;ofi_rxm. Intel MPI and Open MPI work with this method, but MPICH fails with ENODATA. If I specifically set FI_PROVIDER="tcp;ofi_rxm" it is fine. Is this expected? Does MPICH check provider names again and fail?

**Comment by hzhou** [comment]
I believe `FI_PROVIDER` is a flag checked directly by `libfabric`; MPICH does not check this flag. It depends on how libfabric is installed. Currently the default in MPICH is use embedded libfabric, which will have the provider statically compiled in -- hence, `ENODATA` if you ask to load provider modules dynamically. To set provider during configure time, you may try: `--with-device=ch4:ofi:tcp,ofi_rxm`. Alternatively, you may specify environment variable (which MPICH checks) `MPIR_CVAR_OFI_USE_PROVIDER=tcp` to influence the runtime provider selection.

If you compile MPICH with an external `libfabric`, and if that is configured with dynamic provider module, I believe `FI_PROVIDER`, along with `FI_PROVIDER_PATH` should work.

I personally have not tested these configurations extensively. Let me know if it works for you.

**Comment by aingerson** [comment]
Thanks for the help, @hzhou 
Sorry I think I misspoke and messed up my different tests. I've been experimenting with the environment variables and this is what I'm finding:
- Setting MPIR_CVAR_OFI_USE_PROVIDER=tcp does not pick up tcp;ofi_rxm
- Setting FI_PROVIDER=tcp does pick up tcp;ofi_rxm

Is this the expected behavior?

**Comment by hzhou** [comment]
`MPIR_CVAR_OFI_USE_PROVIDER` is used by MPICH to specify desired provider within the loaded providers. `FI_PROVIDER` asks `libfabric` to load specific provider modules. Seems you are using a libfabric installation that uses dynamic provider modules; so if you do not specify `FI_PROVIDER`, none of the modules will be loaded. 

I believe your described behavior is expected behavior.

**Comment by aingerson** [comment]
tcp;ofi_rxm is the ofi format for layering a core and utility provider. Tcp is the main protocol being used, but the ofi_rxm utility provider is layered on top to support FI_EP_RDM endpoint types like MPI uses. Utility providers should not get filtered out if FI_PROVIDER is used to select a specific provider. In other words, if I set FI_PROVIDER=tcp and ask for hints->ep_attr->type = FI_EP_RDM, it should return tcp;ofi_rxm as the provider since tcp alone does not support FI_EP_RDM.

I'm not sure I understand what you're describing about dynamic provider modules... I don't think my libfabric has dynamic provider modules. I am building it from source without specifying --enable-ofi-rxm=dl etc. Is that what you mean?
Libfabric uses the FI_PROVIDER variable to prevent loading unneeded providers. If it is not set, it loads all providers for use. 

I would expect that MPIR_CVAR_OFI_USE_PROVIDER would work similarly. I.e. if I set it to tcp, then it should use tcp;ofi_rxm since fi_getinfo returns tcp;ofi_rxm. Instead, somehow it's getting filtered out. If I don't set FI_PROVIDER but set MPIR_CVAR_OFI_USE_PROVIDER, tcp;ofi_rxm, sockets, verbs;ofi_rxm, and psm2 all get selected correctly. But specifying only core providers that need a utility provider (ie tcp or verbs alone) are getting filtered out when the core provider is selected with MPIR_CVAR_OFI_USE_PROVIDER.

**Comment by hzhou** [comment]
@aingerson Thanks for the explanation. I am not familiar how libfabric works. All `FI_xxx` variables are controlled by libfabric, so I am surprised it would behave differently between MPICH, OpenMPI, or IntelMPI. Are you sure you were testing with the same libfabric installation?

As for MPICH side, I just realized that I am not sure which configuration we are discussing. Could you provide your full `configure` lines?

`MPIR_CVAR_OFI_USE_PROVIDER` is simply passed to `fi_getinfo` in the `hint` struct. Supposedly `libfabric` should filter the lists based on the hint. In `ch3`, we just pick the first provider from the list. In `ch4`, We'll try to match the name exactly. If the match failed, we'll fall back to match with capability. I'll do some homework before I can comment on how libfabric internally does the filtering.

**Comment by aingerson** [comment]
If MPIR_CVAR_OFI_USE_PROVIDER just gets passed into hints, then I would guess tcp;ofi_rxm does get returned. I'm using ch4 so if what you say about trying to match the name exactly is true, then I'm guessing that's probably the issue - tcp gets passed into hints, libfabric returns tcp;ofi_rxm, and then ch4 filters out tcp;ofi_rxm because it does not match MPIR_CVAR_OFI_USE_PROVIDER exactly. Do you know why it checks for an exact name match?

Here's my configure line:
./configure --prefix=/home/aingerson/mpich --disable-oshmem --enable-fortran=no --with-device=ch4:ofi --enable-ch4-direct=netmod --with-libfabric=/home/aingerson/workspace/install      /libfabric/head

I should also clarify that I'm a developer for libfabric. We had some irregularities in our CI that I'm working on resolving.

**Comment by hzhou** [comment]
Currently, MPICH ofi code is mostly tested on socket and psm2 providers, mostly the former. I suspect the default capability set is constructed based on the capability of socket provider. My local testing with only ethernet always picks the socket provider even `tcp;ofi_rxm` was the first provider `libfabric` returned. Looking at debug log, `tcp;ofi_rxm` was rejected because it "doesn't support tagged interfaces".

Check for the exact name match is the simplest. Otherwise, we have to fully understand the detailed libfabric provider name systems and cook up some partial name matching magic. Although in this case, if `tcp` to match `tcp;ofi_rxm` is desirable, it can be easily accommodated.

**Comment by aingerson** [comment]
tcp;ofi_rxm definitely supports tagged messages. I'm guessing you're looking at the tcp output alone. Here are my debug logs when I use MPIR_CVAR_OFI_USE_PROVIDER="tcp" 

libfabric:97545:core:core:fi_getinfo_():966<debug> fi_getinfo: provider tcp returned success
...
libfabric:97545:core:core:fi_getinfo_():966<debug> fi_getinfo: provider ofi_rxm returned success

So it looks like it does pick up tcp;ofi_rxm but MPICH is the one that skips it because of the name mismatch. I know we saw a similar issue with Intel MPI and their I_MPI_OFI_PROVIDER variable so we had to temporarily use our FI_PROVIDER variable to select the provider instead. But now it's working. This is how we select the provider in Intel MPI and Open MPI:
I_MPI_OFI_PROVIDER="tcp"
--mca mtl_ofi_provider_include tcp
Both Intel MPI and Open MPI return tcp;ofi_rxm as expected but MPICH does not.

I still don't really understand why MPICH needs to filter the list of providers by name again. I don't think you need any sort of partial name matching added. I think we just need to remove the name matching completely. If MPIR_CVAR_OFI_USE_PROVIDER is getting passed in directly to hints, this should be enough. What is the objective behind MPICH's name matching?

**Comment by wesbland** [comment]
@aingerson @hzhou 

I believe (though I have not tested to confirm) that utility providers probably don't jive well with our capability sets system that we use to make sure that the particular version/provider combination that we get from OFI has all of the needed features. We also use it to ask for the "best" set of features that we know for a particular provider.

One of the potential problems here is that I believe we do some string matching if you ask for a particular provider with `MPIR_CVAR_OFI_USE_PROVIDER` and if you ask for `tcp`, it wouldn't match `tcp;ofi_rxm`. If you leave `MPIR_CVAR_OFI_USE_PROVIDER` empty, it *might* work, but I'm not sure.

Adding support for utility providers is something we should definitely be getting up and running ASAP. We've been dragging along the sockets provider much longer than the OFI folks say we should have been (in addition to all of the other providers that need a utility provider to get the correct set of features to support MPI).

**Comment by aingerson** [comment]
Hmm yeah, sounds like there needs to be a little work done to improve support for utility providers in MPICH. There is no one maintaining sockets so the sooner you can move to tcp the better. For what it's worth, not setting the MPICH provider variable and instead setting FI_PROVIDER="tcp" lets all of our MPI tests pass. Same with setting it to "verbs;ofi_rxm" - though, as noted in #4187 I'm seeing a hang using MLX hardware but not HFI so I'm not sure what's up with that.

Sounds like a capability set needs to be added for ofi_rxd to solve #4185. I'll close the other two issues and rename this one to general ofi utility provider issues to consolidate the comments.

Thanks!

**Comment by hzhou** [comment]
@wesbland I agree that we really need to support the `tcp` provider, since it is more actively developed than `socket` provider.

@aingerson Since we do not do partial name matching, setting `MPIR_CVAR_OFI_USE_PROVIDER` to `tcp` essentially is the same as not setting it. So it is all up to capability matching. The `tcp;ofi_rxm` currently fails at this check:
```
        } else if (MPIDI_OFI_ENABLE_TAGGED &&
                   (!(prov_use->caps & FI_TAGGED) ||
                    !(prov_use->caps & FI_DIRECTED_RECV) ||
                    prov_use->domain_attr->cq_data_size < 4)) {
```
For `tcp;ofi_rxm`, `FI_TAGGED` is set, `cq_data_size` is 8, but `FI_DIRECTED_RECV` is not set, which is the reason we rejected it. 

If we disable the socket provider, then the default capability set will fail to match any provider, and we'll try to match with the minimal set. There is report saying even that doesn't work. I need check for why (sounds like a bug).

**Comment by hzhou** [comment]
@wesbland @raffenet I think our current provider matching code is a bit too complex and not contributing to the ease of user in practice. Here is the suggestion:

* Remove `MPIR_CVAR_OFI_USE_PROVIDER`. If `FI_PROVIDER` works in picking specific provider, then another variable in MPICH name space is redundant and potentially confusing.

* Only match the capability with `MINIMAL` set, hopefully this will result in picking the first provider that `libfabric` returns. Upon return, we always set the runtime capability to the actual capability set supplied in the provider, so it should not resulting any reduction of capability during runtime. For example, even `FI_TAGGED` is not required in the `MINIMAL` set, since the returned provider has `FI_TAGGED`, we will still use tagged interface.

* I think our previous code worries too much on picking the *best* provider. My suggestion is to stop worrying and start trusting `libfabric` will always return the *best* provider at top of their list. @aingerson Is there bases for this trust?

**Comment by hzhou** [comment]
> Sounds like a capability set needs to be added for ofi_rxd to solve #4185. I'll close the other two issues and rename this one to general ofi utility provider issues to consolidate the comments.

I think it is better to leave these as separate open issues until we know exactly the causes. I'll re-open them.

**Comment by hzhou** [comment]
@aingerson This what I found:
* `FI_PROVIDER='tcp;ofi_rxm' mpirun -n 1 ./cpi` works fine.
`libfabric` returns 5 lists, all with the same name `tcp;ofi_rxm`. I don't understand what are the differences of these 5. They will fail the round when MPICH tries to match with "default" capability set, but the first one will get picked in the second round when MPICH tries to match with "minimal" capability set. Once picked, the application runs fine.

* `FI_PROVIDER=tcp mpirun -n 1 ./cpi` fails.
`libfabric` returns 10 list. The first 5 have name `tcp:ofi_rxm`, followed by another 6 (or more) with provider name `tcp`. The first 5 `tcp` providers fails the default capability set matching (all with FI_TAGGED:0, FI_DIRECTED_RECV:0, cq_data_size:0) but the 6th one actually matched and was selected. Once it is selected, MPICH will use the associated hint call `fi_getinfo` again, and it was failed, resulting an abort.

**Comment by aingerson** [comment]
In regards to your previous question about whether libfabric returns the "best" provider, that's sort of right. The problem is that providers will not set capabilities or features if an app did not request them but they do support them. For example, rxm supports FI_TAGGED and FI_DIRECTED_RECV but if you request the minimal support, rxm will not set FI_TAGGED or FI_DIRECTED_RECV because it does not think the app needs it and so it can enable an optimized path with FI_TAGGED or FI_DIRECTED_RECV support. So that could be a source of an issue as well.

In terms of the second question - about the bogus list of providers - I'm not entirely sure I understand what MPICH is doing or which one it is picking but probably what you're seeing is providers with a different address/address format. If you want to see what providers are available on a system, you can run 'fi_info' or 'fi_info -p tcp' for a specific provider and add '-v' for verbose mode.

**Comment by hzhou** [comment]
> The problem is that providers will not set capabilities or features if an app did not request them but they do support them.

Ahh I recall this issue come up before and probably is the source of current code. It makes sense for libfabric and its providers coded flexible. But MPICH also want to be flexible. Flexible + Flexible apparently makes a mess :(

**Comment by aingerson** [comment]
Yeah I imagine this is the source of the issue... I'm not sure if there's an elegant solution. 

In terms of libfabric filtering out capabilities that are not requested and MPICH requesting the minimal set, I think it makes sense to add rxd MPICH settings so that we can let MPICH be flexible for its minimal settings but at the same time optimize for what rxd needs. 
At the same time, there seems to be something off with the minimal settings since it is somehow falling to the RMA path which rxd supports but MPICH needs ordering which rxd doesn't.

For the tcp;ofi_rxm issue, I think it makes sense to drop the MPIR_CVAR_OFI_USE_PROVIDER since FI_PROVIDER will pick the correct provider and the use of both of them is redundant (however, probably keeping the MPICH variable but just have that set FI_PROVIDER for backwards compatibility). 

As for the verbs;ofi_rxm mellanox issue, I think that's a separate issue from the utility provider mess.

**Comment by hzhou** [comment]
@aingerson First an update: there was a bug in MPICH code that results in the wrong provider being selected when none of the providers matches a default capability set. It should go on to try match with minimal capability set, but the bug resulted in picking the last provider. The fix is in PR #4197.

Second, a question: when we call `fi_getinfo` without hint, will the providers set "maximal" capabilities or "minimal" capabilities" or a "default" capability that is somewhere in-between?

**Comment by hzhou** [comment]
Refer to https://github.com/pmodels/mpich/pull/4197#issuecomment-557844163. Moving to `tcp:ofi_rxm` from `sockets` means we no longer can test MPICH's TAGGED and RMA path on machines with only ethernet. That would be a bummer.

**Comment by wesbland** [comment]
> * Remove `MPIR_CVAR_OFI_USE_PROVIDER`. If `FI_PROVIDER` works in picking specific provider, then another variable in MPICH name space is redundant and potentially confusing.

I don't have a problem with this on its own, but I think we might use `MPIR_CVAR_OFI_USE_PROVIDER` to do string matching if the user requests a specific provider.

> * Only match the capability with `MINIMAL` set, hopefully this will result in picking the first provider that `libfabric` returns. Upon return, we always set the runtime capability to the actual capability set supplied in the provider, so it should not resulting any reduction of capability during runtime. For example, even `FI_TAGGED` is not required in the `MINIMAL` set, since the returned provider has `FI_TAGGED`, we will still use tagged interface.

This would cause problems for the original reason we had the capability sets (to be able to remove a bunch of branches at compile time that we don't need if we know what the OFI provider will do).

> * I think our previous code worries too much on picking the _best_ provider. My suggestion is to stop worrying and start trusting `libfabric` will always return the _best_ provider at top of their list. @aingerson Is there bases for this trust?

"I want to believe." I've always seen things I didn't expect when I looked at the list of providers that gets spit out of the `fi_get_info` call.

**Comment by hzhou** [comment]
@wesbland Now that I refactored the provider selection code in PR #4197, I think we can keep the current code logic for now.

**Comment by wesbland** [comment]
I’m probably not going to be able to review this until I’m back from vacation next week.

**Comment by aingerson** [comment]
Thanks @hzhou for the fix. I'll pull in the changes and update you on what I'm seeing.

> Moving to `tcp:ofi_rxm` from `sockets` means we no longer can test MPICH's TAGGED and RMA path on machines with only ethernet. That would be a bummer.
I don't know if I understand why you wouldn't be able to test TAGGED and RMA. tcp;ofi_rxm supports both.

> Second, a question: when we call `fi_getinfo` without hint, will the providers set "maximal" capabilities or "minimal" capabilities" or a "default" capability that is somewhere in-between?
Without hints at all, the provider should return all functionality, ie a "maximal" set of capabilities.

**Comment by aingerson** [comment]
@hzhou I tried your changes and am still getting an error when using the MPICH variable to select the provider but it still works with the FI_PROVIDER variable. Is that the intent of the change? Should we still be using the libfabric environment variable instead of the MPICH one?

**Comment by hzhou** [comment]
@aingerson What is MPICH variable value that you tried?

**Comment by aingerson** [comment]
MPIR_CVAR_OFI_USE_PROVIDER

**Comment by hzhou** [comment]
"tcp" won't match `tcp:ofi_rxm` and the `tcp` (without utility provider) won't meet the minimum provider capability.

**Comment by aingerson** [comment]
So if I want to specify "tcp" as the provider and have it automatically pick up "tcp;ofi_rxm" like Intel MPI and Open MPI do, I still need to use FI_PROVIDER instead of the MPICH internal variable?

**Comment by hzhou** [comment]
Is "tcp" matching "tcp;ofi_rxm" desirable? If user wants wildcard-like behavior, then not to set MPIR_CVAR_OFI_USE_PROVIDER and let MPICH try to match the providers with capability is, in my opinion, the right approach.

**Comment by hzhou** [comment]
> I don't know if I understand why you wouldn't be able to test TAGGED and RMA. tcp;ofi_rxm supports both.

We will follow up on this with this issue: #4200

**Comment by aingerson** [comment]
> EDIT: I would like to note that `MPIR_CVAR_OFI_USE_PROVIDER ` and `FI_PROVIDER` serve different purposes. The former controls capability set, the latter filters the list. If one focuses on netmod capability or feature, use `MPIR_CVAR_OFI_USE_PROVIDER `. If one focuses on testing provider, use the latter. Interested in both, then both can be used together.

Ok, that makes more sense. So in our case, we want to make sure to test the correct provider, so we should definitely use FI_PROVIDER to select the correct provider. Do we lose anything (performance, functionality, or testing scope) by only setting that and not setting `MPIR_CVAR_OFI_USE_PROVIDER`?

**Comment by hzhou** [comment]
> Ok, that makes more sense. So in our case, we want to make sure to test the correct provider, so we should definitely use FI_PROVIDER to select the correct provider. Do we lose anything (performance, functionality, or testing scope) by only setting that and not setting `MPIR_CVAR_OFI_USE_PROVIDER`?

Short answer, no. MPICH will try to use the default capability set (richer set) first before fall back to minimum set.

**Comment by aingerson** [comment]
> Short answer, no. MPICH will try to use the default capability set (richer set) first before fall back to minimum set.

Do you mean default for rxm or default/maximal capabilities in general?

**Comment by hzhou** [comment]
> 
> 
> > Short answer, no. MPICH will try to use the default capability set (richer set) first before fall back to minimum set.
> 
> Do you mean default for rxm or default/maximal capabilities in general?

 default/maximal capabilities in general.

**Comment by aingerson** [comment]
Ok great that should work. Then I think we will just switch our testing to use FI_PROVIDER="tcp" and let MPICH use the best capability set out of the providers. I think this specific issue can be closed then since it seems like we were not using `MPIR_CVAR_OFI_USE_PROVIDER` as intended. But let's track the tcp;ofi_rxm TAGGED and RMA issue with the other issue you opened.

**Comment by hzhou** [comment]
@aingerson This issue shall be closed automatically when #4197 merges.
